pkg/agent/multicluster/stretched_networkpolicy_controller.go

```diff
@@ -0,0 +1,426 @@
+	if err != nil {
+		klog.ErrorS(err, "Pod UpdateFunc can't get the Namespace of a Pod", "name", curPod.Name, "namespace", curPod.Namespace)
+		return
+	}
+	normalizedLabel := multicluster.GetNormalizedLabel(podNS.Labels, curPod.Labels, podNS.Name)
+	classifierFlowUpdate := s.getClassifierFlowUpdate(getPodReference(curPod), normalizedLabel)
+	s.queue.Add(classifierFlowUpdate)
+}
+
+// processPodDelete handles Pod delete events. It deletes the Pod from the
+// labelToPods and podToLabel. After Pod is deleted, its classifier flow will also
+// be deleted by podConfigurator. So no need to enqueue this Pod to update its
+// classifier flow.
+func (s *StretchedNetworkPolicyController) processPodDelete(old interface{}) {
+	oldPod, oldIsPod := old.(*v1.Pod)
+	if !oldIsPod {
+		klog.ErrorS(errors.New("received unexpected object"), "Pod DeleteFunc can't process event", "obj", old)
+		return
+	}
+	oldPodRef := getPodReference(oldPod)
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	podLabel := s.podToLabel[oldPodRef]
+	s.deleteLabelToPod(podLabel, oldPodRef)
+	delete(s.podToLabel, oldPodRef)
+}
+
+// processNamespaceUpdate handles Namespace update events. It only enqueues all
+// Pods in this Namespace if the Labels of this Namespace has been updated.
+func (s *StretchedNetworkPolicyController) processNamespaceUpdate(old, cur interface{}) {
+	oldNS, oldIsNS := old.(*v1.Namespace)
+	curNS, curIsNS := cur.(*v1.Namespace)
+	if !oldIsNS || !curIsNS {
+		klog.ErrorS(errors.New("received unexpected object"), "Namespace UpdateFunc can't process event", "obj", cur)
+		return
+	}
+	if reflect.DeepEqual(oldNS.Labels, curNS.Labels) {
+		klog.V(2).InfoS("Namespace UpdateFunc received a Namespace update event, but labels are the same. Skip it", "namespace", curNS.Name)
+		return
+	}
+	allPodsInNS, err := s.podLister.Pods(curNS.Name).List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "Namespace UpdateFunc can't list Pods in the Namespace", "namespace", curNS.Name)
+	}
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	for _, pod := range allPodsInNS {
+		if pod.Spec.HostNetwork {
+			continue
+		}
+		normalizedLabel := multicluster.GetNormalizedLabel(curNS.Labels, pod.Labels, curNS.Name)
+		classifierFlowUpdate := s.getClassifierFlowUpdateLocked(getPodReference(pod), normalizedLabel)
+		s.queue.Add(classifierFlowUpdate)
+	}
+}
+
+// processLabelIdentityAddOrUpdate handles labelIdentity add or update event. It will
+// 1. Add/Update labelIdentity to labelIdentityCache
+// 2. Enqueue all Pods with this label Identity
+func (s *StretchedNetworkPolicyController) processLabelIdentityAddOrUpdate(cur interface{}) {
+	labelIdentity, isLabelID := cur.(*v1alpha1.LabelIdentity)
+	if !isLabelID {
+		klog.ErrorS(errors.New("received unexpected object"), "LabelIdentity AddOrUpdateFunc can't process event", "obj", cur)
+		return
+	}
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.labelIdentityCache[labelIdentity.Spec.Label] = labelIdentity.Spec.ID
```

Comment:
this could be acheived by adding an index to labelIdentity informer.

Reply:
Do you mean we don't need a cache map here? We can just use labelidentity informer with index to read the ID that we need.

Reply:
Yes, it could maintain the cache and index atomatically and reduce the critical section here.